pkg/syncer/k8scloudoperator/placement.go

```diff
@@ -410,6 +419,204 @@ func stampPVCWithError(ctx context.Context, client kubernetes.Interface,
+		if i == currPVCParentReplica {
+			continue
+		}
+		// Construct the cousin PVC name and add it to the list
+		//cousinPVCName := volume.ObjectMeta.Name + "-" + parentSTS.Name + "-" + strconv.Itoa(i)
+		cousinPVCName := currPVCPrefix + "-" + strconv.Itoa(i)
+		cousinPVCs = append(cousinPVCs, cousinPVCName)
+	}
+
+	return cousinPVCs, nil
+}
+
+// Removes host at the given index
+func filterHost(s []string, index int) []string {
+	return append(s[:index], s[index+1:]...)
+}
+
+func getPVCObject(cousinPVCName string, pvcList []v1.PersistentVolumeClaim) *v1.PersistentVolumeClaim {
+	for _, pvcItem := range pvcList {
+		if pvcItem.Name == cousinPVCName {
+			return &pvcItem
+		}
+	}
+
+	return nil
+}
+
+// eliminateNodesWithPvcOfSiblingReplica filters out the nodes that have bound PVCs of sibling replicas.
+// It finds cousin PVCs of currPVC and elimates all nodes which have at least one cousin PVC placed on them.
+func eliminateNodesWithPvcOfSiblingReplica(ctx context.Context, client kubernetes.Interface, currPVC *v1.PersistentVolumeClaim, candidateHosts []string) ([]string, error) {
+	log := logger.GetLogger(ctx)
+
+	containerOrchestratorUtility := commonco.ContainerOrchestratorUtility
+	if containerOrchestratorUtility == nil {
+		clusterFlavor, err := cnsconfig.GetClusterFlavor(ctx)
+		if err != nil {
+			log.Debugf("Failed retrieving cluster flavor. Error: %v", err)
+			return candidateHosts, nil
+		}
+		containerOrchestratorUtility, err = commonco.GetContainerOrchestratorInterface(ctx, common.Kubernetes, clusterFlavor, *admissionhandler.COInitParams)
+		if err != nil {
+			log.Debugf("failed to get k8s interface. err: %v", err)
+			return candidateHosts, nil
+		}
+	}
+
+	if !containerOrchestratorUtility.IsFSSEnabled(ctx, common.SiblingReplicaBoundPvcCheck) {
+		log.Infof("FSS to check for sibling replica's PVCs is not enabled.")
+		return candidateHosts, nil
+	}
+
+	// Proceed only if it is a vDPP PVC
+	pvcLabels := currPVC.GetLabels()
+	appplatformLabelVal, labelOk := pvcLabels[appplatformLabel]
+	if !labelOk {
+		log.Infof("Not a vDPP PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	// Do not proceed if it is a standalone pod
+	pvcAnnotation := currPVC.ObjectMeta.Annotations
+	if _, ok := pvcAnnotation[nodeAffinityAnnotationKey]; ok {
+		log.Infof("Node affinity key already specified for PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	cousinPVCs, err := getCousinPVCs(ctx, client, currPVC)
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	selectLabel := fmt.Sprintf("%s=%s", appplatformLabel, appplatformLabelVal)
+	allPVCs, err := client.CoreV1().PersistentVolumeClaims(currPVC.Namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: selectLabel,
+	})
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	for _, cousinPVC := range cousinPVCs {
+		pvc := getPVCObject(cousinPVC, allPVCs.Items)
```

Comment:
How about constructing a map out of allPVCs instead of doing linear search every time?